addons/package_hierarchy/models/package_links.py

```diff
@@ -0,0 +1,185 @@
+    depth = fields.Integer(compute="_compute_would_be_depth")
+
+    def construct(self):
+        for parent_package, links in self.groupby("parent_id"):
+            links.mapped("child_id").write(
+                {"parent_id": parent_package.id if parent_package else False}
+            )
+
+    @api.model
+    def create_top_level_unlinks(self, packages, move_lines=False):
+        # Warm the cache
+        packages.get_move_lines_of_children(aux_domain=[("id", "in", move_lines.ids)])
+        link_vals = []
+        # Create unlinks from higher level packages
+        for top_level_pack in packages:
+            if top_level_pack.parent_id:
+                mls = top_level_pack.get_move_lines_of_children(aux_domain=[("id", "in", move_lines.ids)])
+                link_vals.append(
+                    {
+                        "parent_id": False,
+                        "child_id": top_level_pack.id,
+                        "move_line_ids": [(6, 0, mls.ids if mls else False)],
+                    }
+                )
+
+        if link_vals:
+            return self.create(link_vals)
+        return self.browse()
+
+    @api.depends("parent_id", "child_id")
+    def _compute_name(self):
+        for link in self.filtered(lambda p: not isinstance(p.id, models.NewId)):
+            link.name = (
+                "Link %s and %s" % (link.parent_id.name, link.child_id.name)
+                if link.parent_id and link.child_id
+                else "Unlink parent of %s" % link.child_id.name
+            )
+
+    @api.constrains("parent_id", "child_id")
+    def constrain_links(self):
+        """Check links against constriants.
+        These need to be checked against associated links as they may interact.
+        NOTE: To be 'more correct' we should check against all links from all moves
+        that are in progress, however, this may be computationally expensive for large
+        systems.
+        """
+        # Get all links that are associated with the same move and/or moveline:
+        move_lines = self.mapped("move_line_ids")  # Direct move lines
+        # Add the move lines of associated moves
+        move_lines |= move_lines.mapped("move_id").mapped("move_line_ids")
+        links = move_lines.mapped("u_result_package_link_ids")
+        # If some links are not associated with movelines, we still want to check them:
+        links |= self
+        links._validate_links()
+
+    def _validate_links(self):
+        """Validate package links to ensure that no constraints are broken."""
+        max_package_depth = self.env.user.get_user_warehouse().u_max_package_depth
+
+        # Sanitize links, check for repeated children as we should not have any
+        # Only exception may be moving a package from one package into another package. This will
+        # create a 'un-link' (no-parent) and a 'link' (with parent).
+        children_count = {}
+        for link in self:
```

Comment:
Can `groupby('child_id')` replace this loop ? This way we don't need the filter in line 98.